[MXNET-258] Improve flaky test_random.test_shuffle

asitstands opened the pull request:
## Description ##

This sets a random seed for `test_random.test_shuffle`. As discussed in #10277, the test requires too long running time to be both strict and not flaky. Fixing the seed for the random numbers would be a practical workaround assuming that the random number sequence by the seed is not special.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

marcoabreu commented on tests/python/unittest/test_random.py:
```diff
@@ -554,7 +554,7 @@ def compute_expected_prob():
     mx.test_utils.assert_almost_equal(exp_cnt_true.asnumpy(), exp_cnt[true_classes].asnumpy(), rtol=1e-1, atol=1e-2)
 
-@with_seed()
+@with_seed(1)
```
Could you add a comment with a link to the issue in order to let other developers know why we're fixing the seed?

reminisce replied:
Have you tried seed that makes the test fail? seed=1 seems too special.

asitstands replied:
@reminisce I tried the failed seed but the test passed. It could be from the difference of the environment. The failure was on Windows but I don't have a Windows machine. Seed 1 is not special at all. If it is special, that means the random number generator is very poor. There must be no special seed for good random number generators. More specifically 1 is not special for the generator `mt19937` used on CPU and `philox4_32_10` used on GPU.

asitstands replied:
@marcoabreu I added a comment for the fixed seed.

marcoabreu replied:
Aaah I remember, the device id which makes the entire fixed seeding on GPUs pointless :/ Is the device id a sequential number (e.g. 0-7 for GPU0-7) and thus consistent in between environments or is it some type of unique or arbitrary ID?

Could you try submitting a commit with the failing seed? Ideally, the test should always fail.

asitstands replied:
@marcoabreu I submitted the commit with the failed seed. The test fails again on Windows but is ok on linux (http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-10313/4/pipeline/592). The failed test is on CPU. I cannot find the test result on Windows/GPU from the log. The test might exited before `test_operator_gpu` run. GPU uses different kind of random number generator so it could succeed on Windows with the same seed. I'm not sure why the same generator `mt19937` produces different random sequence in Windows and linux. Maybe VC++ and gcc has different implementation of `mt19937`.

asitstands replied:
@marcoabreu I figured out why it succeed in linux but fails in Windows with the same seed. It is simply `shuffle` uses different implementations for linux and Windows. For linux it uses GNU's parallel version of `std::random_shuffle` and for windows it uses sequential `std::shuffle`. The two are based on totally different algorithms.

marcoabreu replied:
Aaah this makes sense, thanks for diving deep on this! I guess the Windows version is a *lot* slower than the parallelized GNU version, right? Is this something we can change easily or is this out of our control and better not to change?

Seems like this test definitely brought us some knowledge. What do you think about adding a comment to the random generators documentation, describing that two different implementation are being used, depending on the platform? Also, it'd be great if we could modify that assert to not only fail but also print the distribution in relation to the threshold - right now, it just fails and I don't know by how much the distribution is off.

marcoabreu commented on tests/python/unittest/test_random.py:
```diff
@@ -554,7 +554,8 @@ def compute_expected_prob():
     mx.test_utils.assert_almost_equal(exp_cnt_true.asnumpy(), exp_cnt[true_classes].asnumpy(), rtol=1e-1, atol=1e-2)
 
-@with_seed()
+# See issue #10277 for the fixed seed.
```
Please add a link to ensure that developers are able to resolve the reference in future - e.g. when we change bug trackers

asitstands replied:
I added the link.

asitstands commented:
In the last commit, I increased the number of samples and removed the fixed seed. The assertion is also updated to print the difference from the expectation if it fails.
